pymc/logprob/abstract.py

```diff
@@ -169,3 +162,49 @@ def __init__(self, scalar_op, *args, **kwargs):
                 f"Acceptable types are {self.valid_scalar_types}"
             )
         super().__init__(scalar_op, *args, **kwargs)
+
+    def __str__(self):
+        return f"Measurable{super().__str__()}"
+
+
+class ValuedRV(Op):
+    r"""Represents the association of a measurable variable and its value.
+
+    A `ValuedVariable` node represents the pair :math:`(Y, y)`, where
+    :math:`Y` is a random variable and :math:`y \sim Y`.
+
+    Log-probability (densities) are functions over these pairs, which makes
+    these nodes in a graph an intermediate form that serves to construct a
+    log-probability from a model graph.
+    """
+
+    def make_node(self, rv, value):
+        assert isinstance(rv, Variable)
+        assert isinstance(value, Variable)
+        return Apply(self, [rv, value], [rv.type(name=rv.name)])
+
+    def perform(self, node, inputs, out):
+        raise NotImplementedError("ValuedVar should not be present in the final graph!")
+
+    def infer_shape(self, fgraph, node, input_shapes):
+        return [input_shapes[0]]
+
+
+valued_rv = ValuedRV()
+
+
+class PromisedValuedRV(Op):
+    r"""Marks a variable as being promised a valued variable in the logprob method."""
```

Comment:
Revisiting this comment after skimming a bit more

I now see that there are a set of newly defined early measurable ir rewrites and I imagine that these promised value RVs are temporarily substituted into that intermediate graph before applying the other logprob rewrites. Why does this approach prevent breaking interdependencies?

Reply:
We're trying to save the same problem of not breaking dependencies. The funny thing with Join/MakeVector is that they combine multiple RVs, potentially interdependent, into a single composite valued node. Only in the logp function is this value split and sent to each component, but we still want to prevent the same outer-problem that motivated this PR. If you have:

```python
x1 = pt.random.normal() * 5
x2 = pt.random.normal(x1 * 8)
xs = pt.stack([x1, x2])
xs_vv = pt.vector("xs_vv", shape=(2,))
```

By the time you get to the logp of `xs_vv`, you want to basically do the same thing that the new test does, plus concatenate the logp terms. To prevent the conditioning from breaking during the IR rewrites we would want to introduce the ValuedRV nodes. We have to do this as an early rewrite, before anything comes up and breaks it. In normal situations we also do it before any rewrites in the `construct_ir_fgraph` code, but for join/stack this already accounts as some sort of inference.

Now why Promised and not just vanilla Valued? Just for convenience, because we still want a function from `xs_vv` to `stack([logp(x1), logp(x2)])` in the end, and if I split the values, I wouldn't know how to stack them later in the loop in `conditional_logp` (or if this was a single `logp` call). So we just "promise" there will be values to avoid rewrites from breaking the dependency that will be required in the logp function.

Another way of thinking is that we are basically trying to truncate the graphs between valued nodes and do manipulations only within these subgraphs, not across them. We could have literally split the graph, done IR in each and collect logp expressions. What we are doing is identical. Either way, we would still need to do a further split for the RVs within Join/MakeVector.